lib/Frontend/FrontendInputsAndOutputs.cpp

```diff
@@ -379,3 +380,17 @@ bool FrontendInputsAndOutputs::hasDependencyTrackerPath() const {
   return hasDependenciesPath() || hasReferenceDependenciesPath() ||
          hasLoadedModuleTracePath();
 }
+
+PrimarySpecificPaths
+FrontendInputsAndOutputs::getPSPsForAtMostOnePrimary() const {
```

Comment:
I don't know if it really makes sense to have a *mental* model of WMO as "zero primaries". That is, this strikes me as a weird function name. What are you actually trying to do?

Reply:
There are cases where we won't support batch mode with > 1 primary. For example, when compiling a SILModule instead of a source file. In such cases, we can get the PSPs without knowing the InputFile for which code is being generated. That's what this function is doing: getting the primary-specific paths when there are not a multiplicity of primaries.

Here is the implementation:

    return PrimarySpecificPaths(        
                              getSingleOutputFilename(),
                              hasInputsProducingMainOutputs()
                                  ? firstInputProducingOutput().file()
                                  : StringRef(),
                              supplementaryOutputs());

Basically it gets the same information that would have been used before batch mode.